### ISSUE TITLE
System.InvalidCastException thrown when parsing some torrent files

### ISSUE THREAD

**Issue by FrankChen021** [Krusen/BencodeNET#39]
````
var parser = new BencodeParser(); 
var torrent = parser.Parse<Torrent>(torrentFile);
````

i use above codes to parse torrent files. It works ok for most files, but fails for some while there's an exception thrown from `Parse` method as below

````
System.InvalidCastException: Not all elements are of type 'BencodeNET.Objects.BList'.  ---> 
System.InvalidCastException: Unable to cast object of type 'BencodeNET.Objects.BString' to type 'BencodeNET.Objects.BList'.    
at System.Linq.Enumerable.CastIterator[TResult](IEnumerable source)+MoveNext()   
at System.Collections.Generic.List`1..ctor(IEnumerable`1 collection)   
at System.Linq.Enumerable.ToList[TSource](IEnumerable`1 source)    
at BencodeNET.Objects.BList`1..ctor(IEnumerable`1 objects)   
at BencodeNET.Objects.BList.AsType[T]()    --- End of inner exception stack trace ---    
at BencodeNET.Objects.BList.AsType[T]()    
at BencodeNET.Parsing.TorrentParser.ParseTrackers(BDictionary data, Encoding encoding)    at BencodeNET.Parsing.TorrentParser.CreateTorrent(BDictionary data)   
at BencodeNET.Parsing.TorrentParser.Parse(BencodeStream stream)    
at BencodeNET.Parsing.BencodeParser.Parse[T](Stream stream)   
at BencodeNET.Parsing.BencodeParser.Parse[T](String filePath)    
at
````

i've tried convert those failed torrent files to magnetic links on some sites( eg: http://www.torrent2magnet.com/) , there's no exception reported by those sites.

the version of BencodeNET i used is version 2.3.0.

the attachment is a torrent file which causes the exception: 

[Bumblebee 2018 BluRay 1080p AVC Atmos TrueHD7.1 (2).torrent.zip](https://github.com/Krusen/BencodeNET/files/3693511/Bumblebee.2018.BluRay.1080p.AVC.Atmos.TrueHD7.1.2.torrent.zip)

**Comment by Krusen** [comment]
This happens because some torrent files do not follow the specification 100% and has the `announce-list` field as strings instead of lists of strings according to the spec (http://bittorrent.org/beps/bep_0012.html).

Thanks for reporting this. I'll have it fixed in a future version (when using `TorrentParserMode.Tolerant`).

For now you should be able to get arround this issue by adding a `FixedTorrentParser` class and replace the original `TorrentParser`, like this:

```csharp
var bencodeParser = new BencodeParser();
bencodeParser.Parsers.AddOrReplace<Torrent>(new FixedTorrentParser(bencodeParser));
var torrent = bencodeParser.Parse<Torrent>(torrentFile);


// Torrent parser with fix for announce list as strings
public class FixedTorrentParser : TorrentParser
{
    public FixedTorrentParser(IBencodeParser parser)
        : base(parser, TorrentParserMode.Tolerant)
    {

    }

    protected override IList<IList<string>> ParseTrackers(BDictionary data, Encoding encoding)
    {
        // Specification: http://bittorrent.org/beps/bep_0012.html

        var trackerList = new List<IList<string>>();
        var primary = new List<string>();
        trackerList.Add(primary);

        // Get single 'announce' url and add it to the primary list if there is any
        var announce = data.Get<BString>(TorrentFields.Announce)?.ToString(encoding);
        if (!string.IsNullOrEmpty(announce))
        {
            primary.Add(announce);
        }

        // Get the 'announce-list' list's
        var announceLists = data.Get<BList>(TorrentFields.AnnounceList);
        if (announceLists == null)
            return trackerList;

        // According to the specification it should be a list of lists
        if (announceLists.All(x => x is BList))
        {
            var lists = announceLists.AsType<BList>() as IList<BList>;
            if (lists.Any())
            {
                // Add the first list to the primary list and remove duplicates
                primary.AddRange(lists.First().AsStrings(encoding));
                trackerList[0] = primary.Distinct().ToList();

                // Add the other lists to the lists of lists of announce urls
                trackerList.AddRange(lists.Skip(1).Select(x => x.AsStrings(encoding).ToList()));
            }
        }
        // It's not following the specification, it's strings instead of lists
        else if (ParseMode == TorrentParserMode.Tolerant && announceLists.All(x => x is BString))
        {
            // Add them all to the first list
            primary.AddRange(announceLists.AsStrings(encoding));
            trackerList[0] = primary.Distinct().ToList();
        }

        return trackerList;
    }
}
```

**Comment by FrankChen021** [comment]
thanks for your solution. it works!

**Comment by Krusen** [comment]
Fixed for 3.0
